packages/beacon-node/src/chain/blocks/verifyBlocksExecutionPayloads.ts

```diff
@@ -41,14 +43,102 @@ export async function verifyBlocksExecutionPayload(
+  //  parent of any block imported in forkchoice is post-merge and currently we could only
+  //  have blocks[0]'s parent imported in the chain as this is no longer one by one verify +
+  //  import.)
+  //
+  // For a block with SYNCING status (called optimistic block), it's okay to import with
+  // SYNCING status as EL could switch into syncing
+  //
+  // 1. On intial startup/restart
+  // 2. When some reorg might have occured and EL doesn't has a parent root
+  //    (observed on devnets)
+  // 3. Because of some unavailable (and potentially invalid) root but there is no way
+  //    of knowing if this is invalid/unavailable. For unavailable block, some proposer
+  //    will (sooner or later) build on the available parent head which will
+  //    eventually win in fork-choice as other validators vote on VALID blocks.
+  //
+  // Once EL catches up again and respond VALID, the fork choice will be updated which
+  // will either validate or prune invalid blocks
+  //
+  // When to import such blocks:
+  // From: https://github.com/ethereum/consensus-specs/pull/2844
+  // A block MUST NOT be optimistically imported, unless either of the following
+  // conditions are met:
+  //
+  // 1. Parent of the block has execution
+  //
+  //     Since with the sync optimizations, the previous block might not have been in the
+  //     forkChoice yet, so the below check could fail for safeSlotsToImportOptimistically
+  //
+  //     Luckily, we can depend on the preState0 to see if we are already post merge w.r.t
+  //     the blocks we are importing.
+  //
+  //     Or in other words if
+  //       - block status is syncing
+  //       - and we are not in a post merge world and is parent is not optimistically safe
+  //       - and we are syncing close to the chain head i.e. clock slot
+  //       - and parent is optimistically safe
+  //
+  //      then throw error
+  //
+  //
+  //       - if we are are not yet in post merge world (preState0's isMergeTransitionComplete
+  //         is false) and
+  //       - if we haven't yet imported a post merge ancestor in forkchoice i.e.
+  //       - and we are syncing close to the clockSlot, i.e. merge Transition could be underway
+  //
+  //
+  // 2. The current slot (as per the system clock) is at least
+  //    SAFE_SLOTS_TO_IMPORT_OPTIMISTICALLY ahead of the slot of the block being
+  //    imported.
+  //    This means that the merge transition could be underway and we can't afford to import
+  //    a block which is not fully validated as it could affect liveliness of the network.
+  //
+
+  // For this segment of blocks:
+  //   We are optimistically safe with respec to this entire block segment if:
+  //    - we have transitioned to post-merge world or
+  //    - all the blocks are way behind the current slot
+  //    - or we have already imported a post-merge parent of first block of this chain in forkchoice
+  const lastBlock = blocks[blocks.length - 1];
```

Comment:
Does this function ensure that there are more than 0 blocks?

Reply:
may be somewhere from where its called? because it didn't give any type issues, but will return early to be safe

Reply:
added a throw just to be safe